nautobot_ssot/jobs/base.py

```diff
@@ -33,31 +36,135 @@
 """
 
 
-class DataSyncBaseJob(BaseJob):
+class DataSyncBaseJob(BaseJob):  # pylint: disable=too-many-instance-attributes
     """Common base class for data synchronization jobs.
 
     Works mostly as per the BaseJob API, with the following changes:
 
-    - Concrete subclasses are responsible for implementing `self.sync_data()`, **not** `self.run()`.
+    - Concrete subclasses are responsible for implementing `self.sync_data()` (or related hooks), **not** `self.run()`.
     - Subclasses may optionally define any Meta field supported by Jobs, as well as the following:
       - `dry_run_default` - defaults to True if unspecified
       - `data_source` and `data_target` as labels (by default, will use the `name` and/or "Nautobot" as appropriate)
       - `data_source_icon` and `data_target_icon`
     """
 
     dry_run = BooleanVar()
+    memory_profiling = BooleanVar()
+
+    def load_source_adapter(self):
+        """Method to instantiate and load the SOURCE adapter into `self.source_adapter`.
+
+        Relevant available instance attributes include:
+
+        - self.kwargs     (corresponds to the Job's `data` input, including 'dry_run' option)
+        - self.job_result (as per Job API)
+        """
+        raise NotImplementedError
+
+    def load_target_adapter(self):
+        """Method to instantiate and load the TARGET adapter into `self.target_adapter`.
+
+        Relevant available instance attributes include:
+
+        - self.kwargs     (corresponds to the Job's `data` input, including 'dry_run' option)
+        - self.job_result (as per Job API)
+        """
+        raise NotImplementedError
+
+    def calculate_diff(self):
+        """Method to calculate the difference from SOURCE to TARGET adapter and store in `self.diff`.
+
+        This is a generic implementation that you could overwrite completely in you custom logic.
+        """
+        if self.source_adapter and self.target_adapter:
+            self.diffsync_flags = (
```

Comment:
I think it would be good to check if `diffsync_flags` is already defined first, this way it's possible to provide a different set of flags without overwriting the entire method

Reply:
sure, I will move the definition to init, so anyone would be able to overwrite them at any time

Reply:
Possibly make this a class-level attribute `cls.default_diffsync_flags` or such?